microscope/cameras/ximea.py

```diff
@@ -0,0 +1,161 @@
+}
+
+#trig types from define file....
+# XI_TRG_SOURCE = {
+#     "XI_TRG_OFF": c_uint(0),    #Camera works in free run mode.
+#     "XI_TRG_EDGE_RISING": c_uint(1),    #External trigger (rising edge).
+#     "XI_TRG_EDGE_FALLING": c_uint(2),    #External trigger (falling edge).
+#     "XI_TRG_SOFTWARE": c_uint(3),    #Software(manual) trigger.
+#     "XI_TRG_LEVEL_HIGH": c_uint(4),    #Specifies that the trigger is considered valid as long as the level of the source signal is high.
+#     "XI_TRG_LEVEL_LOW": c_uint(5),    #Specifies that the trigger is considered valid as long as the level of the source signal is low.
+#    }
+
+
+@Pyro4.expose
+@Pyro4.behavior('single')
+class XimeaCamera(devices.CameraDevice):
+    def __init__(self, *args, **kwargs):
+        super(XimeaCamera, self).__init__(**kwargs)
+        self._acquiring = False
+        self._exposure_time = 0.1
+        self._triggered = False
+
+    def _fetch_data(self):
+        if self._acquiring and self._triggered:
+            self.handle.get_image(self.img)
+            self._logger.info('Sending image')
+            self._triggered = False
+            data = np.array(list(self.img.get_image_data_raw()),dtype=np.uint16).reshape(self.img.width,self.img.height)
+            return data
+
+    def abort(self):
+        self._logger.info('Disabling acquisition.')
+        if self._acquiring:
+            self._acquiring = False
+        self.handle.stop_acquisition()
+
+    def initialize(self):
+        """Initialise the camera.
+
+        Open the connection, connect properties and populate settings dict.
+        """
+        try:
+            self.handle = xiapi.Camera()
+            self.handle.open_device()
+        except:
+            raise Exception("Problem opening camera.")
+        if self.handle == None:
+            raise Exception("No camera opened.")
+        self._logger.info('Initializing.')
+        self.img=xiapi.Image()
+
+    def make_safe(self):
+        if self._acquiring:
+            self.abort()
+
+    def _on_disable(self):
+        self.abort()
+
+    def _on_enable(self):
+        self._logger.info("Preparing for acquisition.")
+        if self._acquiring:
+            self.abort()
+        self._acquiring = True
+        #actually start camera
+        self.handle.start_acquisition()
+        self._logger.info("Acquisition enabled.")
+        return True
+
+    def set_exposure_time(self, value):
+        #exposure times are set in us.
+        self.handle.set_exposure(int(value*1.0E6))
+
+    def get_exposure_time(self):
+        #exposure times are in us, so multiple by 1E-6 to get seconds.
+        return (self.handle.get_exposure()*1.0E-6) 
+
+    def get_cycle_time(self):
+        return (self.handle.get_exposure()*1.0E-6)
+
+    def _get_sensor_shape(self):
+        return (self.img.width,self.image.height)
+
+
+    def get_trigger_type(self):
+        trig=self.handle.get_trigger_source()
+        return devices.TRIGGER_SOFT
```

Comment:
This always returns TRIGGER_SOFT, regardless of the actual trigger setting. 
Does get_trigger_source return one of the XI_TRG_ constants, defined in the comments above? If so, these should be mapped to trigger modes/types that microscope recognises. I think David added a new mix-in for this, maybe?

Reply:
I found the TriggerTargetMixIn, but could use some advice on the intended usage.

Reply:
> I found the TriggerTargetMixIn, but could use some advice on the intended usage.

One needs to set the private `_trigger_type` and `_trigger_mode` properties during construction. The value in this properties should be an instance of the `TriggerType` and `TriggerMode` enums. Then one needs to define the `set_trigger` abstract method with whatever implementation is suitable to the device.

Reply:
I think this was a quick hack that got left in. The hardware trigger code does work and is being used on DeepSIM. This function should return a proper value of course!

Reply:
OK - so does set_trigger have to set the trigger on the hardware, and set _trigger_type and _trigger_mode appropriately? If so, couldn't we just pass a mapping to the constructor?

Reply:
> so does set_trigger have to set the trigger on the hardware, and set _trigger_type and _trigger_mode appropriately? If so, couldn't we just pass a mapping to the constructor?

Not necessarily setting on the hardware. The alpao mirror does not have a setting on the hardware to trigger mode. One needs to remember what mode was selected and code appropriately when queueing patterns.

I'm guessing with mapping, you mean a map from microscope trigger enum to the sdk trigger enum.  However, the sdk may support them without enums which is what happens in the alpao mirror. And some triggers triggers may support only a subset of trigger modes.

Basically, implementation is completely left up to the concrete class because of so many special cases.

Reply:
Right - good points. Got it, I think. Where there is a clear n-to-1 relationship between device and microscope trigger modes, then we can just put a mapping on the device class, I guess.